openedx/core/djangoapps/programs/utils.py

```diff
@@ -127,6 +127,43 @@ def engaged_programs(self):
 
         return programs
 
+    @cached_property
+    def failed_course_runs(self):
+        """
+        Determine which course runs have been failed by the user.
+
+        Returns:
+            list of strings, each representing a certificate course run key
+        """
+        course_run_certificates = certificate_api.get_certificates_for_user(self.user.username)
+        return [
+            unicode(certificate['course_key']) for certificate in course_run_certificates
+            if not certificate_api.is_passing_status(certificate['status'])
+        ]
+
+    def does_course_still_qualify_as_in_progress(self, now, enrolled_run_modes, course):
+        # Part 1: Check if any of the seats you are enrolled in qualify this course as in progress
+        enrolled_runs = [run for run in course['course_runs'] if run['is_enrolled']]
+        # Check if you are enrolled in the required mode for the run
+        runs_with_required_mode = [
+            run for run in enrolled_runs
+            if run['type'] in enrolled_run_modes[run['key']]
+        ]
+        if runs_with_required_mode:
+            # Check if the runs you are enrolled in with the right mode are not failed
+            not_failed_runs = [run for run in runs_with_required_mode if run not in self.failed_course_runs]
+            if not_failed_runs:
+                return True
+        # Part 2: Check if any of the seats you are not enrolled in
+        # in the runs you are enrolled in qualify this course as in progress
+        upgrade_deadlines = [
+            seat['upgrade_deadline'] for run in enrolled_runs for seat in run['seats']
+            if seat['type'] == run['type'] and run['type'] not in enrolled_run_modes[run['key']]]
```

Comment:
This is way too big to be a comprehension! Please use a for loop to make this readable.